src/app/qgsdockablewidgethelper.cpp

```diff
@@ -0,0 +1,258 @@
+    return;
+
+  if ( docked )
+  {
+    // going from window -> dock
+    mDock = new QgsDockWidget( mOwnerWindow );
+    mDock->setWindowTitle( mWindowTitle );
+    mDock->setWidget( mWidget );
+    setupDockWidget();
+
+    connect( mDock, &QgsDockWidget::closed, [ = ]()
+    {
+      mDockGeometry = mDock->geometry();
+      mIsDockFloating = mDock->isFloating();
+      mDockArea = mOwnerWindow->dockWidgetArea( mDock );
+      emit closed();
+    } );
+  }
+  else
+  {
+    // going from dock -> window
+    mDialog = new QDialog( mOwnerWindow, Qt::Window );
+
+    mDialog->setWindowTitle( mWindowTitle );
+    QVBoxLayout *vl = new QVBoxLayout();
+    vl->setContentsMargins( 0, 0, 0, 0 );
+    vl->addWidget( mWidget );
+    mDialog->setLayout( vl );
+    mDialog->raise();
+    mDialog->show();
+
+    connect( mDialog, &QDialog::finished, [ = ]()
+    {
+      mDialogGeometry = mDialog->geometry();
+      emit closed();
+    } );
+
+    mDialog->setGeometry( mDialogGeometry );
+  }
+}
+
+void QgsDockableWidgetHelper::setWindowTitle( const QString &title )
+{
+  mWindowTitle = title;
+  if ( mDialog )
+  {
+    mDialog->setWindowTitle( title );
+  }
+  if ( mDock )
+  {
+    mDock->setWindowTitle( title );
+  }
+}
+
+void QgsDockableWidgetHelper::setDialogGeometry( const QRect &geom )
+{
+  mDialogGeometry = geom;
+  if ( !mDialog )
+    return;
+  mDialog->setGeometry( geom );
+}
+
+void QgsDockableWidgetHelper::setDockGeometry( const QRect &geom, bool isFloating, Qt::DockWidgetArea area )
+{
+  mDockGeometry = geom;
+  mIsDockFloating = isFloating;
+  mDockArea = area;
+
+  setupDockWidget();
+}
+
+QRect QgsDockableWidgetHelper::dialogGeometry() const
+{
+  if ( mDialog )
+    return mDialog->geometry();
+  return mDialogGeometry;
+}
+
+QRect QgsDockableWidgetHelper::dockGeometry() const
+{
+  if ( mDock )
+    return mDock->geometry();
+  return mDockGeometry;
+}
+
+bool QgsDockableWidgetHelper::isDockFloating() const
+{
+  if ( mDock )
+    return mDock->isFloating();
+  return mIsDockFloating;
+}
+
+Qt::DockWidgetArea QgsDockableWidgetHelper::dockFloatingArea() const
+{
+  if ( mDock )
+    return mOwnerWindow->dockWidgetArea( mDock );
+  return mDockArea;
+}
+
+void QgsDockableWidgetHelper::setupDockWidget()
+{
+  if ( !mDock )
+    return;
+  mDock->setFloating( mIsDockFloating );
+  if ( mDockGeometry.isEmpty() )
+  {
+    // try to guess a nice initial placement for view - about 3/4 along, half way down
+    mDock->setGeometry( QRect( static_cast< int >( mWidget->rect().width() * 0.75 ), static_cast< int >( mWidget->rect().height() * 0.5 ), 400, 400 ) );
+    mOwnerWindow->addDockWidget( mDockArea, mDock );
+  }
+  else
+  {
+    if ( !mIsDockFloating )
+    {
+      // ugly hack, but only way to set dock size correctly for Qt < 5.6
+      mDock->setFixedSize( mDockGeometry.size() );
+      mOwnerWindow->addDockWidget( mDockArea, mDock );
+      mDock->resize( mDockGeometry.size() );
+      QgsApplication::processEvents(); // required!
```

Comment:
Please explain in code comments why this processEvents() call is required - we don't like to see this call randomly in code, it needs justification :-)

Reply:
I just copied this from https://github.com/qgis/QGIS/blob/f64b7863a4a2c0dc263bf470cc2f0ea3d31cd73b/src/app/qgisapp.cpp#L4758 to keep things consistent with existing code